src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java

```diff
@@ -79,54 +79,49 @@ public GitService() {
      * @throws IOException
      * @throws InterruptedException
      */
-    public Repository getOrCheckoutRepository(Participation participation) throws IOException, InterruptedException {
+    public Repository getOrCheckoutRepository(Participation participation) throws IOException, InterruptedException, GitAPIException {
         URL repoUrl = participation.getRepositoryUrlAsUrl();
-        Repository repository = getOrCheckoutRepository(repoUrl);
+        Repository repository = getOrCheckoutRepository(repoUrl, true);
         repository.setParticipation(participation);
         return repository;
     }
 
     /**
      * Get the local repository for a given remote repository URL. If the local repo does not exist yet, it will be checked out.
      *
-     * @param repoUrl The remote repository.
+     * @param repoUrl   The remote repository.
+     * @param pullOnGet Pull from the remote on the checked out repository, if it does not need to be cloned.
      * @return
      * @throws IOException
      * @throws InterruptedException
      */
-    public Repository getOrCheckoutRepository(URL repoUrl) throws IOException, InterruptedException {
+    public Repository getOrCheckoutRepository(URL repoUrl, boolean pullOnGet) throws IOException, InterruptedException, GitAPIException {
 
         Path localPath = new java.io.File(REPO_CLONE_PATH + folderNameForRepositoryUrl(repoUrl)).toPath();
 
-        // check if Repository object already created and available in cachedRepositories
-        if (cachedRepositories.containsKey(localPath)) {
-            // in this case we pull for changes to make sure the Git repo is up to date
-            Repository repository = cachedRepositories.get(localPath);
-            // disable auto garbage collection because it can lead to problems
-            repository.getConfig().setString("gc", null, "auto", "0");
-
-            pull(repository);
+        // First try to just retrieve the git repository from our server, as it might already be checked out.
+        try {
+            Repository repository = getRepositoryByLocalPath(localPath);
+            if (pullOnGet) {
+                pull(repository);
+            }
             return repository;
         }
-
-        // make sure that multiple clone operations for the same repository cannot happen at the same time
-        int numberOfAttempts = 5;
-        while (cloneInProgressOperations.containsKey(localPath)) {
-            log.warn("Clone is already in progress. This will lead to an error. Wait for a second");
-            Thread.sleep(1000);
-            if (numberOfAttempts == 0) {
-                throw new GitException("Cannot clone the same repository multiple times");
-            }
-            else {
-                numberOfAttempts--;
+        // If the git repository can't be found on our server, clone it from the remote.
+        catch (IOException ex) {
```

Comment:
I don't really like that we work with an exception here for a quite common case that the repository cannot be found. This will then occur for all repository the first time. Can we not just handle this by checking if repository is null?

Reply:
You are right, this is not very nice. As proposed I changed to return null if the repo does not exist.